src/driver.ts

```diff
@@ -45,38 +47,100 @@ function emitAllStatusCodes() {
   return true;
 }
 
-function resolveModule(fromModule: string, toModule: string): string {
-  if (!toModule.startsWith('.')) {
-    return toModule;
-  }
+function defaultResolve() {
+  return undefined;
+}
 
-  const p = path.relative(path.dirname(fromModule), toModule);
-  if (p[0] === '.') {
-    return p;
+export function localResolve(ref: string) {
+  if (ref[0] === '#') {
+    return { name: refToTypeName(ref) };
   }
-  return './' + p;
+  return;
+}
+
+export function compose(...fns: types.Resolve[]): types.Resolve {
+  return (ref, options) => {
+    for (const f of fns) {
+      const match = f(ref, options);
+      if (match) {
+        return match;
+      }
+    }
+    return;
+  };
+}
+
+function makeModuleName(filename: string): string {
+  const parts = path
+    .basename(filename)
+    .replace(/\.[^.]*$/, '')
+    .split(/[^a-zA-Z0-9]/);
+  return [parts[0], ...parts.slice(1).map(capitalize)].join('');
+}
+
+export function generateFile(): types.Resolve {
+  return (ref: string, options: types.Options) => {
+    if (ref[0] === '#') {
+      return;
+    }
+    const localName = ref.replace(/^[^#]+/, '');
+    const fileName = ref.replace(/#.+$/, '');
+    const ymlFile = path.resolve(path.dirname(options.sourceFile), fileName);
+    const generatedFileName = `${path.basename(ymlFile).replace(/\.[^.]*$/, '')}.types.generated`;
+    const moduleName = makeModuleName(fileName);
+    const generatedFile = './' + path.dirname(options.targetFile) + '/' + generatedFileName;
+
+    return {
+      importAs: moduleName,
+      importFrom: generatedFile,
+      name: refToTypeName(localName),
+      generate: () => {
+        generate({
+          openapiFilePath: ymlFile,
+          header: options.header,
+          generatedValueClassFile: generatedFile + '.ts',
+          resolve: options.resolve,
+          externalOpenApiSpecs: options.externalOpenApiSpecs,
+          externalOpenApiImports: options.externalOpenApiImports,
+          emitStatusCode: options.emitStatusCode,
+          unsupportedFeatures: options.unsupportedFeatures
+        });
+      }
+    };
+  };
 }
 
 export function generate(driver: Driver) {
   const file = fs.readFileSync(driver.openapiFilePath, 'utf8');
   const spec: oas.OpenAPIObject = yaml.load(file);
+
+  types.deprecated(
+    driver.generatedServerFile && driver.generatedClientFile,
+    'generating both server and client files from the same definition is frowned upon and will be prevented later on'
```

Comment:
the client and server types might need to diverge when/if we get to support readonly/writeonly properties or @khugo s differing generated types for queries